overlord/snapshotstate/snapshotstate_test.go

```diff
@@ -0,0 +1,662 @@
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error {
+		c.Assert(f(&backend.Reader{
+			// wanted
+			Snapshot: client.Snapshot{SetID: setID, Snap: "a-snap"},
+			File:     shotfile,
+		}), check.IsNil)
+
+		return errBad
+	}
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	snaps, files, err := snapshotstate.SnapNamesInSnapshotSet(setID, nil)
+	c.Assert(err, check.Equals, errBad)
+	c.Check(snaps, check.IsNil)
+	c.Check(files, check.IsNil)
+}
+
+func (snapshotSuite) TestSnapnamesInSnapshotSetNotFound(c *check.C) {
+	setID := uint64(42)
+	shotfile, err := os.Create(filepath.Join(c.MkDir(), "foo.zip"))
+	c.Assert(err, check.IsNil)
+	defer shotfile.Close()
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error {
+		c.Assert(f(&backend.Reader{
+			// not wanted
+			Snapshot: client.Snapshot{SetID: setID - 1, Snap: "a-snap"},
+			File:     shotfile,
+		}), check.IsNil)
+
+		return nil
+	}
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	snaps, files, err := snapshotstate.SnapNamesInSnapshotSet(setID, nil)
+	c.Assert(err, check.Equals, client.ErrSnapshotSetNotFound)
+	c.Check(snaps, check.IsNil)
+	c.Check(files, check.IsNil)
+}
+
+func (snapshotSuite) TestSnapnamesInSnapshotSetEmptyNotFound(c *check.C) {
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error { return nil }
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	snaps, files, err := snapshotstate.SnapNamesInSnapshotSet(42, nil)
+	c.Assert(err, check.Equals, client.ErrSnapshotSetNotFound)
+	c.Check(snaps, check.IsNil)
+	c.Check(files, check.IsNil)
+}
+
+func (snapshotSuite) TestSnapnamesInSnapshotSetSnapNotFound(c *check.C) {
+	setID := uint64(42)
+	shotfile, err := os.Create(filepath.Join(c.MkDir(), "foo.zip"))
+	c.Assert(err, check.IsNil)
+	defer shotfile.Close()
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error {
+		c.Assert(f(&backend.Reader{
+			// not wanted
+			Snapshot: client.Snapshot{SetID: setID, Snap: "a-snap"},
+			File:     shotfile,
+		}), check.IsNil)
+
+		return nil
+	}
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	snaps, files, err := snapshotstate.SnapNamesInSnapshotSet(setID, []string{"b-snap"})
+	c.Assert(err, check.Equals, client.ErrSnapshotSnapsNotFound)
+	c.Check(snaps, check.IsNil)
+	c.Check(files, check.IsNil)
+}
+
+func (snapshotSuite) TestCheckConflict(c *check.C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+	chg := st.NewChange("some-change", "...")
+	tsk := st.NewTask("some-task", "...")
+	tsk.SetStatus(state.DoingStatus)
+	chg.AddTask(tsk)
+
+	// no snapshot state
+	err := snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-change")
+	c.Assert(err, check.ErrorMatches, "internal error: task \"some-task\" of change 1 is missing snapshot information")
+
+	// wrong snapshot state
+	tsk.Set("snapshot", "hello")
+	err = snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-change")
+	c.Assert(err, check.ErrorMatches, "internal error.*")
```

Comment:
because the error text in this case feels too implementation-specific (and might be worth fixing):
`internal error: could not unmarshal state entry "snapshot": json: cannot unmarshal string into Go value of type snapshotstate.snapshotState`

But maybe there's a middle ground. I'll make it match `"internal error.*could not unmarshal .*"`.